File: app/lib/models/message.dart
/// A free-text + photo message — a squad post or a thread reply
/// (specs/api/messages.md). body is nullable (photo-only allowed).
class Message {
  const Message({
    required this.id,
    this.senderName,
    this.senderPhoto,
    this.body,
    this.attachments = const [],
    this.threadCount = 0,
    this.createdAt,
  });

  final String id;
  final String? senderName;
  final String? senderPhoto;
  final String? body;
  final List<MessageAttachment> attachments;
  final int threadCount;
  final DateTime? createdAt;

  factory Message.fromJson(Map<String, dynamic> json) {
    final sender = json['sender'] as Map<String, dynamic>?;
    return Message(
      id: json['id'] as String,
      senderName: sender?['first_name'] as String?,
      senderPhoto: sender?['photo'] as String?,
      body: json['body'] as String?,
      attachments: ((json['attachments'] as List<dynamic>?) ?? const [])
          .map((e) => MessageAttachment.fromJson(e as Map<String, dynamic>))
          .toList(),
      threadCount: (json['thread_count'] as num?)?.toInt() ?? 0,
      createdAt: json['created_at'] == null
          ? null
          : DateTime.tryParse(json['created_at'] as String),
    );
  }
}

/// An image attachment on a message: storage key + public URL (specs/api/uploads.md).
class MessageAttachment {
  const MessageAttachment({required this.key, required this.url});

  final String key;
  final String url;

  factory MessageAttachment.fromJson(Map<String, dynamic> json) =>
      MessageAttachment(
        key: json['key'] as String? ?? '',
        url: json['url'] as String? ?? '',
      );

  Map<String, dynamic> toJson() => {'key': key, 'url': url};
}

File: server/src/db/schema/message.ts
import { pgTable, pgEnum, uuid, text, timestamp, jsonb } from 'drizzle-orm/pg-core'

import { profile } from './identity.ts'
import { squad } from './squad.ts'

// An image attachment on a message: the storage key + its public URL
// (specs/api/uploads.md). Stored as a jsonb array on the message.
export type MessageAttachment = { key: string; url: string }

// Polymorphic thread root kinds (specs/data-model.md). community_event arrives with
// the communities stage.
export const threadTargetType = pgEnum('thread_target_type', [
  'activity',
  'community_event',
  'message',
])

// Free-text + photo posts — covers squad top-level messages AND thread replies
// (one primitive). Exactly one context: squad (top-level squad post) XOR thread_of
// (thread reply), enforced in the domain layer. Attachments arrive with storage.
export const message = pgTable('message', {
  id: uuid('id').primaryKey().defaultRandom(),
  senderId: uuid('sender_id')
    .notNull()
    .references(() => profile.id, { onDelete: 'cascade' }),
  body: text('body'), // nullable ⇒ photo-only (body or attachments must be present)
  // image attachments (object-store keys + public URLs); defaults to empty.
  attachments: jsonb('attachments').$type<MessageAttachment[]>().notNull().default([]),
  // top-level squad message:
  squadId: uuid('squad_id').references(() => squad.id, { onDelete: 'cascade' }),
  // thread reply (polymorphic root):
  threadTargetType: threadTargetType('thread_target_type'),
  threadTargetId: uuid('thread_target_id'),
  createdAt: timestamp('created_at', { withTimezone: true }).notNull().defaultNow(),
})
